### PR TITLE
fix(store): enforce unique table names across types

### PR DESCRIPTION
**Pull request by yonadaaa**
fixes https://github.com/latticexyz/mud/issues/2710

### CODE REVIEW

File: packages/store/src/StoreCore.sol
Comment (holic) [inline packages/store/src/StoreCore.sol:219]: Rather than adding a new table, I think we can just extend this check to look for `ResourceId` with the same name but different type.

Comment (yonadaaa) [reply packages/store/src/StoreCore.sol:219]: How else would you find a `ResourceId` with the same name? There is no global array of all Resource IDs we can traverse

Comment (yonadaaa) [reply packages/store/src/StoreCore.sol:219]: Like, just given a `bytes30` name, how do you find any/all corresponding Resource ID's unless there is a table that specifically indexes that

Comment (yonadaaa) [reply packages/store/src/StoreCore.sol:219]: I see, we can change the type of Table ID to check if there is either an offchain or regular table instead

File: packages/store/src/StoreCore.sol
Comment (holic) [inline packages/store/src/StoreCore.sol:230]: how much worse on gas would this be if we did something like

```solidity
{
  ResourceId onchainTableId = ResourceIdLib.encode(RESOURCE_TABLE, tableId.getNameStore());
  ResourceId offchainTableId = ResourceIdLib.encode(RESOURCE_OFFCHAIN_TABLE, tableId.getNameStore());
  if (ResourceIds._getExists(onchainTableId) || ResourceIds._getExists(offchainTableId)) {
    revert IStoreErrors.Store_TableAlreadyExists(tableId, string(abi.encodePacked(tableId)));
  }
}
```

File: packages/store/src/ResourceId.sol
Comment (holic) [inline packages/store/src/ResourceId.sol:54]: what about `getResourceName`?

File: packages/store/src/StoreCore.sol
Comment (yonadaaa) [reply packages/store/src/StoreCore.sol:230]: It's actually cheaper! I guess because we do two `_getExists` checks regardless.

Much clearer too.

File: packages/store/src/ResourceId.sol
Comment (yonadaaa) [reply packages/store/src/ResourceId.sol:54]: That's kinda better but the World name is also a "resource name" 🤔

File: .changeset/strong-lobsters-shake.md
Comment (holic) [inline .changeset/strong-lobsters-shake.md:5]:
```suggestion
Added a check to `registerTable` that prevents registering both an offchain and onchain table with the same name, making indexing with decoded tables easier.
```

File: .changeset/strong-lobsters-shake.md
Comment (holic) [inline .changeset/strong-lobsters-shake.md:2]: need to bump store/world protocol versions

Comment (yonadaaa) [reply .changeset/strong-lobsters-shake.md:2]: done

**Review by holic — DISMISSED** [review]
looks like need to regenerate docs but otherwise lgtm!

**Comment by holic** [comment]
waiting for @alvrs sign off too, since this is a protocol bump

**Review by alvrs — APPROVED** [review]

**Review by holic — APPROVED** [review]

**Pull request merged**
